src/v/raft/coordinated_recovery_throttle.h

```diff
@@ -0,0 +1,131 @@
+ */
+#pragma once
+
+#include "config/configuration.h"
+#include "seastarx.h"
+#include "utils/token_bucket.h"
+
+#include <seastar/core/sharded.hh>
+
+namespace raft {
+
+/// A recovery throttle that coordinates the total available rate across shards.
+/// Coordination is done by shard 0. The intent of the coordination is to
+/// saturate the available node recovery rate (bandwidth). This is achieved by
+/// redistributing the unused bandwidth from idle shards among the busy shards.
+/// Internally it wraps a simple token bucket (semaphore) that is periodically
+/// reset with the desired capacity as deemed by the coordinator.
+
+/// In order to guarantee fairness, each shard always has access to its fair
+/// share of bandwidth (= total_available/num_shards). Any unused portion of
+/// this fair share is redistributed among the busy shards.
+class coordinated_recovery_throttle
+  : public ss::peering_sharded_service<coordinated_recovery_throttle> {
+public:
+    explicit coordinated_recovery_throttle(
+      config::binding<size_t> /* node capacity in bytes per sec*/,
+      config::binding<bool> /* use static rate allocation*/);
+    coordinated_recovery_throttle(const coordinated_recovery_throttle&)
+      = delete;
+    coordinated_recovery_throttle&
+    operator=(const coordinated_recovery_throttle&)
+      = delete;
+    coordinated_recovery_throttle(coordinated_recovery_throttle&&) = delete;
+    coordinated_recovery_throttle& operator=(coordinated_recovery_throttle&&)
+      = delete;
+    ~coordinated_recovery_throttle() noexcept = default;
+
+    void shutdown();
+
+    ss::future<> start();
+    ss::future<> stop();
+
+    /// A test helper to step through the ticks.
+    ss::future<> tick_for_testing() { return do_coordinate_tick(); }
+
+    ss::future<> throttle(size_t size, ss::abort_source& as) {
+        return _throttler.throttle(size, as);
+    }
+
+    size_t available() const { return _throttler.available(); }
+    size_t waiting_bytes() const { return _throttler.waiting_bytes(); }
+    size_t admitted_bytes() const { return _throttler.admitted_bytes(); }
+
+private:
+    using clock_t = ss::lowres_clock;
+    static constexpr ss::shard_id _coordinator_shard = ss::shard_id{0};
+
+    /// A wrapper around a sempahore that tracks additional metrics
+    /// used for coordination. The capacity of the bucket is refilled
+    /// by the recovery throttle periodically as a part of coordination.
+    class token_bucket {
+    public:
+        explicit token_bucket(size_t /*initial_size*/);
+        ss::future<> throttle(size_t /*bytes*/, ss::abort_source&);
+        /// Resets to the newly passed capacity argument. If the current
+        /// availability is higher, the capacity is reduced, otherwise
+        /// the delta is added.
+        void reset_capacity(size_t /*new capacity*/);
+        void shutdown() { return _sem.broken(); }
+        inline size_t waiting_bytes() const { return _waiting_bytes; }
```

Comment:
inline is implied when its in the header

Reply:
Fixed in next rev.